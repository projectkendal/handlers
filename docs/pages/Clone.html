<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="Page with examples of how Kendal framework can be used">
  <meta name="author" content="Konrad Gancarz & Arkadiusz Ryszewski">

  <title>Kendal - Manual</title>

  <link rel="stylesheet" href="../resources/default.min.css">
  <!-- Bootstrap core CSS -->
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">

  <!-- Bootstrap core JavaScript -->
  <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js" integrity="sha384-ApNbgh9B+Y1QKtv3Rn7W3mgPxhU9K/ScQsAP7hUibX39j7fakFPskvXusvfa0b4Q" crossorigin="anonymous"></script>
  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js" integrity="sha384-JZR6Spejh4U02d8jOt6vLEHfe/JQGiRRSQQxSfFWpi1MquVdAyjUar5+76PVCmYl" crossorigin="anonymous"></script>

  <!-- Highlight.js -->
  <script src="../resources/highlight.min.js"></script>

  <script>hljs.initHighlightingOnLoad();</script>
</head>

<body>

<nav class="navbar navbar-expand-md navbar-dark bg-dark">
  <a class="navbar-brand" href="../Manual.html">Kendal</a>

  <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#menu-navbar" aria-controls="menu-navbar" aria-expanded="false" aria-label="Toggle navigation">
    <span class="navbar-toggler-icon"></span>
  </button>

  <div class="collapse navbar-collapse" id="menu-navbar">
    <ul class="navbar-nav mr-auto">
      <li class="nav-item">
        <a class="nav-link" href="TsFields.html">TypeScript Fields</a>
      </li>
      <li class="nav-item">
        <a class="nav-link" href="StringInterpolation.html">String Interpolation</a>
      </li>
      <li class="nav-item active">
        <a class="nav-link disabled">Clone</a>
      </li>
      <li class="nav-item">
        <a class="nav-link" href="InheritAttributeAndAttrReference.html">Inherit, Attribute & AttrReference</a>
      </li>
    </ul>
    </div>
</nav>

<div class="container" style="margin-top: 10px">
  <p>
    Applied to a method creates new method with name from <code>methodName</code> attribute and list of parameters same as original method.
    Generated method calls the original method, passes the result to an instance of transformer specified with <code>transformer</code> attribute
    and returns transformed value. If <code>methodName</code> is not specified, clone with default name is created.
    Default name is: <code>originalName + "Clone"</code>.
    Return type of generated method is extracted from declaration of the transformer class.
  </p>

    <h4>Vanilla Java</h4>
    <pre><code class="java">
        class SomeClass {

            List&lt;Object&gt; method1(int oneParam, String anotherParam) {
                List&lt;Object&gt; result = new ArrayList();
                // some logic
                return result;
            }

            List&lt;Object&gt; method2(int oneParam, String anotherParam) {
                List&lt;Object&gt; result = new ArrayList();
                // some logic
                return result;
            }

            String method1Size(int oneParam, String anotherParam) {
                return new ListToSizeTransformer().transform(method1(oneParam, anotherParam));
            }

            String method2Size(int oneParam, String anotherParam) {
                return new ListToSizeTransformer().transform(method2(oneParam, anotherParam));
            }

        }

        class ListToSizeTransformer {
            String transform(List&lt;Object&gt; parametersList) {
                return String.valueOf(new ArrayList().size());
            }
        }
    </code></pre>

    <h4>Kendal</h4>
    <pre><code class="java">
        class SomeClass {

            @Clone(transformer = ListToSizeTransformer.class, methodName="method1Size")
            List&lt;Object&gt; method1(int oneParam, String anotherParam) {
                List&lt;Object&gt; result = new ArrayList();
                // some logic
                return result;
            }

            @Clone(transformer = ListToSizeTransformer.class, methodName="method2Size")
            List&lt;Object&gt; method2(int oneParam, String anotherParam) {
                List&lt;Object&gt; result = new ArrayList();
                // some logic
                return result;
            }

        }

        class ListToSizeTransformer implements Clone.Transformer&lt;List&lt;Object&gt;, String&gt; {
            @Override
            public String transform(List&lt;Object&gt; parametersList) {
                return String.valueOf(parametersList.size());
            }
        }
    </code></pre>
</div><!-- /.container -->

</body>
</html>
