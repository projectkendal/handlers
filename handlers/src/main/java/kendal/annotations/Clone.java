package kendal.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

/**
 * Applied to a method creates new method with name from {@link #methodName()} attribute and list of parameters same as original method.
 * Generated method calls the original method, passes the result to an instance of {@link #transformer()}
 * and returns transformed value. If {@link #methodName()} is not specified, clone with default name is created.
 * Default name is: originalName + "Clone"
 * Return type of generated method is extracted from declaration of {@link #transformer()} class.
 */
@Target(ElementType.METHOD)
public @interface Clone {

    Class<? extends Transformer> transformer();

    /**
     * Java does not allow putting {@link java.lang.annotation.Annotation}, which is the base type for all annotations, as annotation member.
     * Let's just use this field as if it was defined.
     * We remove onMethod parameter from AST after processing @Clone usage.
     * /
//    Annotation[] onMethod() default {};

    /**
     * Name of the generated method.
     * Defaults to original method name + "Clone" suffix
     * */
    String methodName() default "";

    interface Transformer<T,R> {
        R transform(T input);
    }
}
