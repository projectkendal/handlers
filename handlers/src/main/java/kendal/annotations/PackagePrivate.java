package kendal.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

@Target(value= ElementType.PARAMETER)
public @interface PackagePrivate {

    boolean makeFinal() default true;
}