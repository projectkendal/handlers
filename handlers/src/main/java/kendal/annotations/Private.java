package kendal.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

@Target(value= ElementType.PARAMETER)
public @interface Private {
}