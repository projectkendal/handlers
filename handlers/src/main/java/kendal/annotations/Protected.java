package kendal.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

@Target({ElementType.PARAMETER, ElementType.ANNOTATION_TYPE})
public @interface Protected {

    boolean makeFinal() default true;
}