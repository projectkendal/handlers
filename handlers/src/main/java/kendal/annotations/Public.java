package kendal.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

@Target(value= ElementType.PARAMETER)
public @interface Public {

    boolean makeFinal() default true;
}