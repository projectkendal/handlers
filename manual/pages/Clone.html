<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="description" content="Page with examples of how Kendal framework can be used">
    <meta name="author" content="Konrad Gancarz & Arkadiusz Ryszewski">

    <title>Kendal - Manual</title>

    <!-- Bootstrap core CSS -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-alpha.6/css/bootstrap.min.css" integrity="sha384-rwoIResjU2yc3z8GV/NPeZWAv56rSmLldC3R/AZzGRnGxQQKnKkoFVhFQhNUwEyJ" crossorigin="anonymous">

    <!-- Highlight.js -->
    <link rel="stylesheet" href="../resources/default.min.css">
    <script src="../resources/highlight.min.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
</head>

<body>

<nav class="navbar navbar-toggleable-md navbar-inverse bg-inverse">
    <a class="navbar-brand" href="../Manual.html">Kendal</a>

    <div class="collapse navbar-collapse">
        <ul class="navbar-nav mr-auto">
            <li class="nav-item">
                <a class="nav-link" href="TsFields.html">TypeScript Fields</a>
            </li>
            <li class="nav-item">
                <a class="nav-link" href="StringInterpolation.html">String Interpolation</a>
            </li>
            <li class="nav-item active">
                <a class="nav-link disabled">Clone</a>
            </li>
            <li class="nav-item">
              <a class="nav-link" href="InheritAttributeAndAttrReference.html">Inherit, Attribute & AttrReference</a>
            </li>
        </ul>
    </div>
</nav>

<div class="container" style="margin-top: 10px">
    <h3>Vanilla Java</h3>
    <pre><code class="java">
        class SomeClass {

            List&lt;Object1&gt; method1(int oneParam, String anotherParam) {
                List&lt;Object1&gt; result = new ArrayList();
                // some logic
                return result
            }

            List&lt;Object2&gt; method2(int oneParam, String anotherParam) {
                List&lt;Object2&gt; result = new ArrayList();
                // some logic
                return result
            }

            String method1Size(int oneParam, String anotherParam) {
                return new ListToSizeTransformer().transform(method1(oneParam, anotherParam));
            }

            String method2Size(int oneParam, String anotherParam) {
                return new ListToSizeTransformer().transform(method2(oneParam, anotherParam));
            }

        }

        class ListToSizeTransformer {
            String transform(List&lt;Object&gt; parametersList) {
                return String.valueOf(new ArrayList().size());
            }
        }
    </code></pre>

    <h3>Kendal</h3>
    <pre><code class="java">
        class SomeClass {

            @Clone(transformer = ListToSizeTransformer.class, methodName="method1Size")
            List&lt;Object1&gt; method1(int oneParam, String anotherParam) {
                List&lt;Object1&gt; result = new ArrayList();
                // some logic
                return result;
            }

            @Clone(transformer = ListToSizeTransformer.class, methodName="method2Size")
            List&lt;Object2&gt; method2(int oneParam, String anotherParam) {
                List&lt;Object2&gt; result = new ArrayList();
                // some logic
                return result;
            }

        }

        class ListToSizeTransformer implements Clone.Transformer&lt;List&lt;Object&gt;, String&gt; {
            @Override
            String transform(List&lt;Object&gt; parametersList) {
                return String.valueOf(parametersList.size());
            }
        }
    </code></pre>
</div><!-- /.container -->


<!-- Bootstrap core JavaScript -->
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-alpha.6/js/bootstrap.min.js" integrity="sha384-vBWWzlZJ8ea9aCX4pEW3rVHjgjt7zpkNpZk+02D9phzyeVkE+jo0ieGizqPLForn" crossorigin="anonymous"></script>
</body>
</html>
