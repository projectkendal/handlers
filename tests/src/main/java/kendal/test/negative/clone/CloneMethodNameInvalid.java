package kendal.test.negative.clone;

import kendal.annotations.Clone;

/*
 * @test
 * @summary clone method name invalid
 * @build TestTransformer
 * @compile/fail/ref=CloneMethodNameInvalid.out CloneMethodNameInvalid.java
 */
@SuppressWarnings("unused")
public class CloneMethodNameInvalid {

    @Clone(transformer = TestTransformer.class, methodName = "!#@%@#%@%#")
    public Integer method() {
        return null;
    }
}
