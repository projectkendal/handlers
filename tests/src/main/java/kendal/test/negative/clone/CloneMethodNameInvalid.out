error: [Kendal] !#@%@#%@%# is not a valid method identifier!
1 error